src/Enforcer.php

```diff
@@ -472,6 +474,188 @@ public function getImplicitUsersForPermission(string ...$permission): array
+     * If there is an obj policy that does not meet the prefix requirement, an ObjConditionException will be thrown.  
+     * 
+     * 1. If the 'objectConditions' array is empty, an EmptyConditionException will be thrown.
+     * This error is thrown because some data adapters' ORM return full table data by default
+     * when they receive an empty condition, which tends to behave contrary to expectations.(e.g. DBALAdapter)
+     * If you are using an adapter that does not behave like this, you can choose to ignore this error.  
+     * 
+     * @param string $user
+     * @param string $action
+     * @param string $prefix
+     * 
+     * @return array
+     * @throws ObjConditionException
+     * @throws EmptyConditionException
+     */
+    public function getAllowedObjectConditions(string $user, string $action, string $prefix): array
+    {
+        $permission = $this->getImplicitPermissionsForUser($user);
+
+        $objectConditions = [];
+        foreach ($permission as $policy) {
+            if ($policy[2] == $action) {
+                if (strpos($policy[1], $prefix) !== 0) {
+                    throw new ObjConditionException('need to meet the prefix required by the object condition');
+                }
+
+                $trimPrefix = function ($string, $prefix) {
+                    if (strpos($string, $prefix) === 0) {
+                        return substr($string, strlen($prefix));
+                    } else {
+                        return $string;
+                    }
+                };
+
+                $objectConditions[] = $trimPrefix($policy[1], $prefix);
+            }
+        }
+
+        if (empty($objectConditions)) {
+            throw new EmptyConditionException('GetAllowedObjectConditions have an empty condition');
+        }
+
+        return $objectConditions;
+    }
+
+    /**
+     * GetImplicitUsersForResource return implicit user based on resource.
+     * For example:  
+     * p, alice, data1, read  
+     * p, bob, data2, write  
+     * p, data2_admin, data2, read  
+     * p, data2_admin, data2, write  
+     * g, alice, data2_admin  
+     * GetImplicitUsersForResource("data2") will return [[bob data2 write] [alice data2 read] [alice data2 write]]  
+     * GetImplicitUsersForResource("data1") will return [[alice data1 read]]  
+     * Note: only users will be returned, roles (2nd arg in "g") will be excluded.  
+     *
+     * @param string $resource
+     *
+     * @return array
+     */
+    public function getImplicitUsersForResource(string $resource): array
+    {
+        $permissions = [];
+        $subIndex = $this->model->getFieldIndex('p', Constants::SUBJECT_INDEX);
+        $objIndex = $this->model->getFieldIndex('p', Constants::OBJECT_INDEX);
+        $rm = $this->getRoleManager();
+
+        $isRole = [];
+        $roles = $this->getAllRoles();
+        foreach ($roles as $role) {
+            $isRole[$role] = true;
+        }
```

Comment:
Please simplify this code:
```suggestion
        $isRole = array_flip($rules);
```

Reply:
That‘s wonderful😲.